Выполнено задание спринта 12.

PUprojects opened the pull request:
В проект внедрена база данных h2.
Реализованы репозитории для доступа к данным.
Реализована логика работы с рейтингами и жанрами фильмов. Реализованы тесты репозиториев.
Хранилища в памяти исключены из проекта.

PUprojects committed:
Выполнено задание спринта 12.
В проект внедрена база данных h2.
Реализованы репозитории для доступа к данным.
Реализована логика работы с рейтингами и жанрами фильмов.
Реализованы тесты репозиториев.
Хранилища в памяти исключены из проекта.

PUprojects committed:
Исправлены ошибки checkstyle (локальный их почему-то не увидел)

PUprojects committed:
Исправлены ошибки checkstyle (локальный их почему-то не увидел)

bismuthD commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,52 @@
+CREATE TABLE IF NOT EXISTS users
```
Можно было бы ввести ограничение на уникальность логина и почты на уровне таблицы в БД

create unique index if not exists USER_EMAIL_UINDEX on USERS (email);
create unique index if not exists USER_LOGIN_UINDEX on USERS (login);

bismuthD commented on src/main/java/ru/yandex/practicum/filmorate/repository/JdbcFilmRepository.java:
```diff
@@ -0,0 +1,141 @@
+        source.addValue("mpa", film.getMpa().getId());
+
+        film.setId(insert(INSERT_QUERY, source));
+        if (film.getGenres() != null)
```
getGenres() может отдать пустую коллекцию, лучше проверять CollectionUtils.isNotEmpty(), там включена уже проверка на null и, что коллекция не empty

bismuthD commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -2,58 +2,91 @@
+                .orElseThrow(() -> new BadRequestException("Not found MPA rating with id " + film.getMpa().getId()));
+        film.setMpa(mpa);
+        if (film.getGenres() != null) {
+            List<Genre> genres = genreRepository.getByIds(film.getGenres().stream().map(Genre::getId).toList());
```
стримы лучше не писать в одну строчку для читабельности - либо выносить в отдельный метод, либо складывать результат в переменную и ее потом использовать. Обычно методы стрим апи через точку пишут с новой строки. Например тут выглядело бы так:

```
List<Genre> genres = genreRepository.getByIds(film.getGenres().stream()
                                         .map(Genre::getId)
                                         .toList());
```

bismuthD commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,52 @@
+(
+    film_id         BIGINT PRIMARY KEY AUTO_INCREMENT,
+    film_name       VARCHAR(255) NOT NULL,
+    description     VARCHAR(255),
```
по ТЗ размер описания - 200 символов

bismuthD reviewed (commented):
Отличная работа, хорошо разобрался со взаимодействием между слоями сервис контроллер репозиторий, оставил несколько комментариев на доработку

PUprojects committed:
Учтены замечания и предложения после ревью:
- добавлено создание уникальных индексов для полей email и login в таблице users;
-добавлена проверка CollectionUtils.isNotEmpty() при здании жанров в FilmService;
- исправлено написание вызова методов stream api;
- установлено ограничение на поле description таблицы films в 200 символов (сгласно ТЗ).

PUprojects committed:
Учтены замечания и предложения после ревью:
- добавлено создание уникальных индексов для полей email и login в таблице users;
-добавлена проверка CollectionUtils.isNotEmpty() при здании жанров в FilmService;
- исправлено написание вызова методов stream api;
- установлено ограничение на поле description таблицы films в 200 символов (сгласно ТЗ).
(и опять checkstyle)
